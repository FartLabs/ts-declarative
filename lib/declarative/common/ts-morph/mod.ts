export * from "./ts-morph.ts";
