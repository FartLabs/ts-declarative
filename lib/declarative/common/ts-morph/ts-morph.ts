import type { Project, SourceFile } from "ts-morph";
import type { Class, Declarative } from "#/lib/declarative/declarative.ts";
import { createDecoratorFactory } from "#/lib/declarative/decorator.ts";

/**
 * TsMorph is a declarative class type from ts-morph.
 */
export interface TsMorph {
  properties: TsMorphProperty[];
}

export interface TsMorphProperty {
  name: string;
  type: string;
  paramIndex?: number;
  // TODO: Add documentation.
  // TODO: Add which class the inherited properties associate with respectively.
}

export interface ValueTsMorph {
  tsMorph?: TsMorph;
}

export function tsMorphDecoratorFactory(
  project: Project,
): (specifier: string | URL) => (target: Class) => Class {
  return createDecoratorFactory({
    initialize: (specifier: URL | string) => {
      const sourceFile = project.getSourceFileOrThrow(specifier.toString());
      return [declarativeTsMorph(sourceFile)];
    },
  });
}

export function declarativeTsMorph<TValue extends ValueTsMorph>(
  sourceFile: SourceFile,
): Declarative<TValue> {
  return (value, name) => {
    return { ...value, tsMorph: getTsMorph(sourceFile, name) } as TValue;
  };
}

export interface ValueType {
  type?: string[];
}

export const type: (
  ...args: Array<string | string[]>
) => (target: Class) => Class = createDecoratorFactory({
  initialize: (...type: Array<string | string[]>) => {
    return [declarativeType(...type)];
  },
});

export function declarativeType<TValue extends ValueType>(
  ...type: Array<string | string[]>
): Declarative<TValue> {
  return <TValue extends ValueType>(value: TValue | undefined): TValue => {
    return { ...value, type: type.flat() } as TValue;
  };
}

function getTsMorph(sourceFile: SourceFile, name: string): TsMorph {
  const classDeclaration = sourceFile.getClass(name);
  const propertyDeclarations = classDeclaration?.getProperties() ?? [];
  const constructorParameterDeclarations =
    classDeclaration?.getConstructors().at(-1)?.getParameters() ?? [];

  return {
    properties: [
      ...propertyDeclarations.map((property): TsMorphProperty => {
        return {
          name: property.getName(),
          type: property.getType().getText(),
        };
      }),
      ...constructorParameterDeclarations.reduce<TsMorphProperty[]>(
        (acc, parameter, i) => {
          if (parameter.getScope() === "public") {
            acc.push({
              name: parameter.getName(),
              type: parameter.getType().getText(),
              paramIndex: i,
            });
          }

          return acc;
        },
        [],
      ),
    ],
  };
}
