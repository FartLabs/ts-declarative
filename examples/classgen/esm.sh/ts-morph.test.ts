import { assertSnapshot } from "@std/testing/snapshot";
import { Project } from "ts-morph";
import { walkEsmSh } from "./ts-morph.ts";

// https://jsr.io/@luca/cases
Deno.test({
  name: "ts-morph project from jsr:@luca/cases",
  permissions: { net: true, read: true, write: true },
  fn: async (t) => {
    const project = new Project({ useInMemoryFileSystem: true });
    await walkEsmSh("jsr:@luca/cases@1.0.0", ({ specifier }, sourceCode) => {
      project.createSourceFile(
        specifier.slice(
          specifier.indexOf("@luca/cases"),
        ),
        sourceCode,
        { overwrite: true },
      );
    });

    await assertSnapshot(
      t,
      project.getSourceFiles().map((sourceFile) => {
        return [sourceFile.getFilePath(), sourceFile.getText()];
      }),
    );
  },
});
