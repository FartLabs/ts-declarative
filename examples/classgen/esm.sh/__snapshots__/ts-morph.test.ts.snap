export const snapshot = {};

snapshot[`ts-morph project from jsr:@luca/cases 1`] = `
[
  [
    "/s",
    'function p(r){let o=/[^\\\\p{Lu}_\\\\-\\\\s]+|\\\\p{Lu}+(?![^\\\\p{Lu}_\\\\-\\\\s])|\\\\p{Lu}[^\\\\p{Lu}_\\\\-\\\\s]*/gu;return Array.from(r.matchAll(o),e=>e[0])}function n(r){return(Array.isArray(r)?r:p(r)).map((e,a)=>a===0?e.toLowerCase():e.toUpperCase()===e?e:e[0].toUpperCase()+e.slice(1).toLowerCase()).join("")}function t(r){return(Array.isArray(r)?r:p(r)).map(e=>e.toLowerCase()).join("_")}function i(r){return(Array.isArray(r)?r:p(r)).map(e=>e.toLowerCase()).join("-")}function c(r){return(Array.isArray(r)?r:p(r)).map(e=>e.toUpperCase()===e?e:e[0].toUpperCase()+e.slice(1).toLowerCase()).join(" ")}function s(r){return(Array.isArray(r)?r:p(r)).map(e=>e.toUpperCase()===e?e:e[0].toUpperCase()+e.slice(1).toLowerCase()).join("")}function u(r){return(Array.isArray(r)?r:p(r)).map(e=>e.toUpperCase()).join("_")}export{n as camelCase,u as constantCase,i as kebabCase,s as pascalCase,t as snakeCase,p as splitPieces,c as titleCase};
//# sourceMappingURL=luca__cases.mjs.map',
  ],
  [
    "/@luca/cases@1.0.0",
    'export * from "/@jsr/luca__cases@1.0.0/denonext/luca__cases.mjs";
',
  ],
]
`;
